### PR TITLE
TopSQL: add table id in resource tag

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve?

Issue Number:  ref https://github.com/pingcap/tidb/issues/55540
Problem Summary:
*: add table id in resource tag
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/kv/kv.go [hunk 1]
```diff
@@ -766,3 +770,87 @@ const (
...
+		}
+		key = k
+		if !key.HasPrefix(tablePrefix) {
+			return 0
+		}
+	}
+	key = key[len(tablePrefix):]
+	_, tableID, _ := codec.DecodeInt(key)
+	return tableID
+}
```
Comment (XuHuaiyu) [inline pkg/kv/kv.go:856]:
```suggestion
var DecodeTableIDFunc func(Key) int64

func decodeTableID(key Key) int64 {
	if DecodeTableIDFunc != nil {
		return DecodeTableIDFunc(key)
	}
	return 0
}
```

``` diff
diff --git a/pkg/tablecodec/tablecodec.go b/pkg/tablecodec/tablecodec.go
index 042f620f2e..698fdfcecc 100644
--- a/pkg/tablecodec/tablecodec.go
+++ b/pkg/tablecodec/tablecodec.go
@@ -1902,3 +1902,7 @@ func VerifyTableIDForRanges(keyRanges *kv.KeyRanges) ([]int64, error) {
        err := keyRanges.ForEachPartitionWithErr(collectFunc)
        return tids, err
 }
+
+func init(){
+       kv.DecodeTableIDFunc = DecodeTableID
+}
```

File: pkg/kv/kv.go [hunk 2]
```diff
@@ -766,3 +770,87 @@ const (
...
+func (b *ResourceGroupTagBuilder) EncodeTagWithKey(key []byte) []byte {
+	tag := &tipb.ResourceGroupTag{}
+	if b.sqlDigest != nil {
+		tag.SqlDigest = b.sqlDigest.Bytes()
+	}
+	if b.planDigest != nil {
+		tag.PlanDigest = b.planDigest.Bytes()
+	}
+	if len(key) > 0 {
+		tag.TableId = decodeTableID(key)
```
Comment (yibin87) [inline pkg/kv/kv.go:815]: Do we need to consider any error cases here?

File: pkg/util/resourcegrouptag/resource_group_tag.go [hunk 3]
```diff
@@ -125,6 +125,12 @@ func GetFirstKeyFromRequest(req *tikvrpc.Request) (firstKey []byte) {
 				}
 			}
 		}
+	case *kvrpcpb.PessimisticLockRequest:
```
Comment (yibin87) [inline pkg/util/resourcegrouptag/resource_group_tag.go:128]: Why add this new case?

File: pkg/kv/kv.go [hunk 4]
```diff
@@ -766,3 +770,87 @@ const (
...
+	accessKey  []byte
+}
+
+// NewResourceGroupTagBuilder creates a new ResourceGroupTagBuilder.
+func NewResourceGroupTagBuilder() *ResourceGroupTagBuilder {
+	return &ResourceGroupTagBuilder{}
+}
+
+// SetSQLDigest sets the sql digest for the request.
+func (b *ResourceGroupTagBuilder) SetSQLDigest(digest *parser.Digest) *ResourceGroupTagBuilder {
```
Comment (yibin87) [inline pkg/kv/kv.go:787]: It seems not common to return self for SetXXX function in tidb's existing go code

File: pkg/kv/kv.go (hunk 2 above)
Comment (nolouch) [reply pkg/kv/kv.go:814]: If the decoded fails, it will be tagged as 0.

File: pkg/kv/kv.go (hunk 4 above)
Comment (nolouch) [reply pkg/kv/kv.go:786]: Here example are: https://github.com/pingcap/tidb/blob/master/pkg/distsql/request_builder.go#L177-L200

File: pkg/util/resourcegrouptag/resource_group_tag.go (hunk 3 above)
Comment (nolouch) [reply pkg/util/resourcegrouptag/resource_group_tag.go:128]: I just think it was missing, so I added it.

File: pkg/kv/kv.go (hunk 4 above)
Comment (yibin87) [reply pkg/kv/kv.go:786]: OK

File: pkg/kv/kv.go (hunk 2 above)
Comment (yibin87) [reply pkg/kv/kv.go:814]: I've no idea, do we need to distinguish real '0' with the error '0'?

Comment (nolouch) [reply pkg/kv/kv.go:814]: I think no need, and the real `table_id` must be greater than 0, https://github.com/pingcap/tidb/blob/master/pkg/meta/meta.go#L253-L254

Comment (yibin87) [reply pkg/kv/kv.go:814]: Got it

**Review by yibin87 — APPROVED** [review]
LGTM

**Comment by nolouch** [comment]
PTAL @XuHuaiyu

**Review by crazycs520 — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**